trieste/models/keras/architectures.py

```diff
@@ -0,0 +1,315 @@
+        elements of the covariance matrix are parametrized. If left as the default `False`,
+        then :class:`~tfp.layers.MultivariateNormalTriL` layer is used where correlations
+        between outputs are learned as well.
+    :return: Keras ensemble model.
+    """
+    input_tensor_spec, output_tensor_spec = get_tensor_spec_from_data(dataset)
+
+    hidden_layer_args = []
+    for i in range(num_hidden_layers):
+        hidden_layer_args.append({"units": units, "activation": activation})
+
+    networks = [
+        GaussianNetwork(
+            input_tensor_spec,
+            output_tensor_spec,
+            hidden_layer_args,
+            independent_normal,
+        )
+        for _ in range(ensemble_size)
+    ]
+    keras_ensemble = KerasEnsemble(networks)
+
+    return keras_ensemble
+
+
+class KerasEnsemble(tf.Module):
+    """
+    This class builds an ensemble of neural networks, using Keras. Individual networks must
+    be instance of :class:`~trieste.models.keras_networks.KerasEnsembleNetwork`. This class
+    is meant to be used with :class:`~trieste.models.keras_networks.DeepEnsemble` model wrapper,
+    which compiles the model.
+    """
+
+    def __init__(
+        self,
+        networks: List[KerasEnsembleNetwork],
+    ) -> None:
+        """
+        :param networks: A list of neural network specifications, one for each member of the
+            ensemble. The ensemble will be built using these specifications.
+        """
+
+        super().__init__()
+
+        for network in networks:
+            if not isinstance(network, KerasEnsembleNetwork):
+                raise ValueError(
+                    f"Individual networks must be an instance of KerasEnsembleNetwork, "
+                    f"received {type(network)} instead."
+                )
+
+        self._networks = networks
+        self._model = self.build_ensemble()
+
+    def __repr__(self) -> str:
+        """"""
+        return f"KerasEnsemble({self._networks!r})"
+
+    @property
+    def model(self) -> tf.keras.Model:
+        """ " Returns built but uncompiled Keras ensemble model."""
+        return self._model
+
+    @property
+    def ensemble_size(self) -> int:
+        """
+        Returns the size of the ensemble, that is, the number of base learners or individual neural
+        network models in the ensemble.
+        """
+        return len(self._networks)
+
+    def build_ensemble(self) -> tf.keras.Model:
+        """
+        Builds the ensemble model by combining all the individual networks in a single Keras model.
+        This method relies on ``connect_layers`` method of :class:`KerasEnsembleNetwork` objects
+        to construct individual networks.
+
+        :return: The Keras model.
+        """
+        inputs = []
+        outputs = []
+        for index, network in enumerate(self._networks):
+            network.network_name = "model_" + str(index) + "_"
```

Comment:
Is it fine to just drop the existing network names rather than incorporating them into the name?

Also, this step arguably belongs earlier, when we set `self._networks` rather than here. If you move it, then you can also simplify this loop into one line (and this function into two):

```suggestion
        inputs, outputs = zip(*[network.connect_layers() for network in self._networks])
```